File: src/main/python/AoC2025_06.py
#! /usr/bin/env python3
#
# Advent of Code 2025 Day 6
#

import sys
from dataclasses import dataclass
from enum import Enum
from enum import auto
from enum import unique
from math import prod
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import CharGrid

TEST = "123 328  51 64 \n 45 64  387 23 \n  6 98  215 314\n*   +   *   +  "


@unique
class Mode(Enum):
    BY_ROWS = auto()
    BY_COLUMNS = auto()


@unique
class Operation(Enum):
    ADD = auto()
    MULTIPLY = auto()

    @classmethod
    def from_string(cls, s: str) -> Self:
        match s:
            case "+":
                return Operation.ADD
            case "*":
                return Operation.MULTIPLY
            case _:
                raise ValueError


@dataclass(frozen=True)
class Problem:
    nums: list[int]
    operation: Operation


@dataclass(frozen=True)
class Worksheet:
    grid: CharGrid
    ops: list[Operation]

    def get_problems(self, mode: Mode) -> list[Problem]:
        problems, ns, j = list[Problem](), list[str](), 0
        for col in self.grid.get_cols_as_strings():
            if col.strip() == "":
                match mode:
                    case Mode.BY_ROWS:
                        rows = CharGrid.from_strings(ns).get_cols_as_strings()
                    case Mode.BY_COLUMNS:
                        rows = CharGrid.from_strings(ns).get_rows_as_strings()
                problems.append(
                    Problem([int(row.strip()) for row in rows], self.ops[j])
                )
                ns = []
                j += 1
            else:
                ns.append(col)
        return problems


Input = Worksheet
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        lines = list(input_data)
        grid = CharGrid.from_strings([s + " " for s in lines[:-1]])
        ops = [Operation.from_string(s) for s in lines[-1].strip().split()]
        return Worksheet(grid, ops)

    def solve(self, worksheet: Worksheet, mode: Mode) -> int:
        def solve_problem(problem: Problem) -> int:
            match problem.operation:
                case Operation.ADD:
                    return sum(problem.nums)
                case Operation.MULTIPLY:
                    return prod(problem.nums)

        return sum(solve_problem(p) for p in worksheet.get_problems(mode))

    def part_1(self, worksheet: Input) -> Output1:
        return self.solve(worksheet, Mode.BY_ROWS)

    def part_2(self, worksheet: Input) -> Output2:
        return self.solve(worksheet, Mode.BY_COLUMNS)

    @aoc_samples(
        (
            ("part_1", TEST, 4277556),
            ("part_2", TEST, 3263827),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2025, 6)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2025_06.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2025_06 extends SolutionBase<AoC2025_06.WorkSheet, Long, Long> {

    private AoC2025_06(final boolean debug) {
        super(debug);
    }

    public static AoC2025_06 create() {
        return new AoC2025_06(false);
    }

    public static AoC2025_06 createDebug() {
        return new AoC2025_06(true);
    }

    @Override
    protected WorkSheet parseInput(final List<String> inputs) {
        return WorkSheet.fromInput(inputs);
    }

    @SuppressWarnings("PMD.TooFewBranchesForSwitch")
    private long solve(final WorkSheet worksheet, final Mode mode) {
        return worksheet.getProblems(mode).stream()
                .mapToLong(
                        p ->
                                switch (p.operation) {
                                    case ADD -> Arrays.stream(p.nums).sum();
                                    case MULTIPLY ->
                                            Arrays.stream(p.nums).reduce(1L, (a, b) -> a * b);
                                })
                .sum();
    }

    @Override
    public Long solvePart1(final WorkSheet worksheet) {
        return this.solve(worksheet, Mode.BY_ROWS);
    }

    @Override
    public Long solvePart2(final WorkSheet worksheet) {
        return this.solve(worksheet, Mode.BY_COLUMNS);
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "4277556"),
        @Sample(method = "part2", input = TEST, expected = "3263827"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST =
            """
            123 328  51 64\s
             45 64  387 23\s
              6 98  215 314
            *   +   *   +\s\s
            """;

    record Problem(long[] nums, Operation operation) {

        enum Operation {
            ADD,
            MULTIPLY;

            public static Operation fromString(final String string) {
                return switch (string) {
                    case "+" -> ADD;
                    case "*" -> MULTIPLY;
                    default -> throw new IllegalArgumentException();
                };
            }
        }

        @Override
        public String toString() {
            final StringBuilder builder = new StringBuilder(100);
            builder.append("Problem [nums=")
                    .append(Arrays.toString(nums))
                    .append(", operation=")
                    .append(operation)
                    .append(']');
            return builder.toString();
        }
    }

    enum Mode {
        BY_ROWS,
        BY_COLUMNS
    }

    record WorkSheet(List<String> strings) {

        public static WorkSheet fromInput(final List<String> inputs) {
            return new WorkSheet(inputs);
        }

        public Set<Problem> getProblems(final Mode mode) {
            final Set<Problem> problems = new HashSet<>();
            final CharGrid grid =
                    CharGrid.from(
                            range(this.strings.size() - 1).stream()
                                    .map(i -> this.strings.get(i) + " ")
                                    .toList());
            final List<Problem.Operation> ops =
                    Arrays.stream(this.strings.getLast().strip().split("\\s+"))
                            .map(Problem.Operation::fromString)
                            .toList();
            final List<String> nums = new ArrayList<>();
            int j = 0;
            for (int col = 0; col < grid.getWidth(); col++) {
                final String s = grid.getColumnAsString(col);
                if (s.isBlank()) {
                    final Stream<String> rows =
                            switch (mode) {
                                case BY_ROWS -> CharGrid.from(nums).getColumnsAsStrings();
                                case BY_COLUMNS -> CharGrid.from(nums).getRowsAsStrings();
                            };
                    problems.add(
                            new Problem(
                                    rows.mapToLong(row -> Long.parseLong(row.strip())).toArray(),
                                    ops.get(j)));
                    nums.clear();
                    j++;
                } else {
                    nums.add(s);
                }
            }
            return problems;
        }
    }
}
